### DISCUSSION TITLE
Variable treated as type alias?

### DISCUSSION THREAD

**Discussion by math-fehr** [microsoft/pyright#3758] [Q&A] [answered]
Hello,
I recently found out that pyright seems to coerce a variable (containing a type) to a type alias.
Here is a small reproducer.

```python
x = int            # pyright indentifies this as a type alias for `Type[int]`
a = type(x)        # pyright identifies this a variable (not a type alias) of type `Type[int]`
b: a = x           # pyright accepts this definition
```
MyPy raises an error on this: https://mypy-play.net/?mypy=latest&python=3.10&gist=a8fba1628b0e481e7f5e68895c5dd1de

Note that `a` cannot be defined as a type alias anyway, since this code will make pyright return an error
```python
from typing import TypeAlias
x = int                       # pyright indentifies this as a type alias for `Type[int]`
a: TypeAlias = type(x)        # Error: type() call should not be used in type annotation
```

Note that it seems to apply for all variables of type `Type[...]`, but fails correctly for this code for instance:
```python
a = 2
b: a = 3 # Error: Illegal type annotation: variable not allowed unless it is a type alias
```

Is this a bug in Pyright, or is this a feature?

**Comment by erictraut** [comment, marked as answer]
In your first sample above, you said that "pyright accepts this definition" on line 3. That's not what I'm seeing.

```python
x = int
a = type(x)
b: a = x # Error: Illegal type annotation: variable not allowed unless it is a type alias
```

This matches the behavior of mypy, which emits the error "Variable "test3.a" is not valid as a type".

Which version of pyright are you using? Please upgrade to the latest version and let me know if you still see the same behavior.

**Reply by erictraut** [reply]
Prior to the introduction `TypeAlias` in [PEP 613](https://peps.python.org/pep-0613/), there was no way for a developer to indicate that a symbol was intended to be a type alias. Type checkers needed to apply various heuristics to "guess" the intent. I've tuned pyright's heuristics over time, and pyright has converged on behavior that largely matches mypy in this regard.

**Reply by math-fehr** [reply]
Indeed, it seems that I had an older version, it seems to be fixed on the latest version (1.1.264).
Thanks a lot!
